tests/backends/base/test_creation.py

```diff
@@ -101,3 +105,38 @@ def test_circular_reference(self):
         obj_ref = ObjectReference.objects.get()
         self.assertEqual(obj.obj_ref, obj_ref)
         self.assertEqual(obj_ref.obj, obj)
+
+    def test_serialize_circular_ref_with_natural_keys(self):
+        # serialize_db_to_string and deserialize_db_from_string handle circular and self-references for models with
+        # natural keys.
```

Comment:
As you mentioned in the first commit `serialize_db_to_string()` doesn't use natural keys so I'm not sure why we mix them with these extra tests :thinking:

Reply:
Well, the bug was that if natural keys *are defined*, serialize_db_to_string would fail, even when it did not even use the natural keys, so this test checks that that now works?

Reply:
Test that I moved to the #12651 works without this patch (with and without natural keys). It failed before 98f23a8af0be7e87535426c5c83058e2682bfdf8.

The remaining part of this test (circular references) was fixed by the first commit d9059b4ad04da685e8eaad9a7b40ef12801f9e9d so we should move it there (we can merge this in advance).

Reply:
Even if this is a single ticket IMO we fixed here multiple small issues with serialization, that's why I would like to handle them in separate bits.

Reply:
Yeah, the self-references were already special-cased, so they did not cause the problem fixed in this PR. I guess I misread your initial comment in this thread to ask why we should bother with testing natural keys at all, but it seems you asked about mixing circular references and self-references, which you now split. So I think this is resolved by moving the test to your new PR?

Reply:
Yes. After merging #12651, I will try to add a remaining case to the first commit and merge it in advance.

Reply:
I moved fix for reloading the test database with circular related objects and natural keys to the #12652.